src/Workspaces/Core/Portable/Shared/Extensions/SyntaxEditorExtensions.cs

```diff
@@ -73,36 +121,40 @@ internal static class SyntaxEditorExtensions
         /// it will call into <paramref name="canReplace"/> to validate if the subsequent fix
         /// can be made or not.
         /// </summary>
-        private static async Task ApplySemanticEditsAsync<TNode>(
+        private static async Task ApplySemanticEditsAsync<TType>(
             this SyntaxEditor editor, Document document,
-            ImmutableArray<TNode> originalNodes,
+            ImmutableArray<TType> originalNodes,
             Func<ISyntaxFactsService, SyntaxNode, SyntaxNode> getSemanticBoundary,
-            Func<SemanticModel, TNode, bool> canReplace,
-            Func<SemanticModel, SyntaxNode, TNode, SyntaxNode> updateRoot,
-            CancellationToken cancellationToken) where TNode : SyntaxNode
+            Func<SemanticModel, TType, SyntaxNode, bool> canReplace,
+            Func<SemanticModel, SyntaxNode, TType, SyntaxNode, SyntaxNode> updateRoot,
+            Func<TType, SyntaxNode> selector,
+            CancellationToken cancellationToken)
         {
+            var originalNodePairs = originalNodes.Select(n => (n, selector(n)));
```

Comment:
can we name our tuple elements, seeing Item1 and Item2 is just confusing below.  Also, consider not using 'nodes' for things that are no longer SyntaxNodes.

Reply:
This should be generic on two types.  We both want it to be strongly typed on a TInput type (i.e. your TType), and the TNode type that 'selector' function (and things like canReplace and updateRoot) operate with.  That will prevent unnecessary casts in the callers.

Reply:
i'm on the fence if you even need to do this.  just calling selector on-demand doesn't seem like a problem.  But it's up to you if you want to make sucha  change.  i'm ok with this as is.